pageserver/src/aux_file.rs

```diff
@@ -0,0 +1,100 @@
+use pageserver_api::key::{Key, AUX_KEY_PREFIX, METADATA_KEY_SIZE};
+use tracing::warn;
+
+/// Create a metadata key from a hash, encoded as [AUX_KEY_PREFIX, 2B directory prefix, first 13B of 128b xxhash].
+fn aux_hash_to_metadata_key(prefix: u16, data: &[u8]) -> [u8; METADATA_KEY_SIZE] {
+    let mut key = [0; METADATA_KEY_SIZE];
+    let hash = twox_hash::xxh3::hash128(data).to_be_bytes();
+    let prefix = prefix.to_be_bytes();
+    key[0] = AUX_KEY_PREFIX;
+    key[1] = prefix[0];
+    key[2] = prefix[1];
+    key[3..16].copy_from_slice(&hash[0..13]);
+    key
+}
+
+/// Encode the aux file into a fixed-size key.
+///
+/// The first byte is the AUX key prefix. We use the next 2 bytes of the key for the directory / aux file type.
+/// We have one-to-one mapping for each of the aux file that we support. We hash the remaining part of the path
+/// (usually a single file name, or several components) into 13-byte hash. The way we determine the 2-byte prefix
+/// is roughly based on the first two components of the path, one unique number for one component.
+///
+/// * pg_logical/mappings -> 0x0101
+/// * pg_logical/snapshots -> 0x0102
+/// * pg_logical/replorigin_checkpoint -> 0x0103
+/// * pg_replslot/ -> 0x0201
+/// * others -> 0xFFFF
+///
+/// If you add new AUX files to this function, please also add a test case to `test_encoding_portable`.
```

Comment:
This comment should be a bit stronger: if someone adds new AUX files to this function then they **must** have never been written before on a neon system.  If someone adds a new aux file "foo/bar", it'll get stored in the 0xffff prefix.  If they later come and edit this function to map "foo/bar" to 0x0304, then that will break existing databases.

In practice that probably means that we will only ever be able to add special cases here under special conditions where we know a key has never been used in postgres on neon before.  That's okay, but it should be made very clear in the comment

Reply:
fixed in https://github.com/neondatabase/neon/pull/7401/commits/6fc849b7442602b5fbd061bb48d34a624889b21d